### PR TITLE
[REF] core: config cleanup (part 2)

### PR DESCRIPTION
**Pull request by Julien00859**
Second part of a comprehensive refactor of the configuration, first part is at https://github.com/odoo/odoo/pull/126099

If you are experiencing problems with this work, can you provide us with the logs and the content of your (maybe multiple) configuration files? **Make sure to remove your admin password!!!**

Task-3380313

### CODE REVIEW

File: odoo/tools/config.py [hunk 1]
```diff
@@ -526,59 +578,48 @@ def die(cond, msg):
...
-        dev_split = [s.strip() for s in opt.dev_mode.split(',')] if opt.dev_mode else []
-        self._runtime_options['dev_mode'] = dev_split + (['reload', 'qweb', 'xml'] if 'all' in dev_split else [])
-
+        for mod in ('base', 'web'):
+            if mod not in self['server_wide_modules']:
+                self._log(logging.INFO, "adding missing %r to %s", mod, self.options_index['server_wide_modules'])
+                self._runtime_options['server_wide_modules'] = [mod] + self['server_wide_modules']
+
+        # ensure default addons-path are present
+        for ad_path in self._default_options['addons_path']:
```
Comment (Julien00859) [inline odoo/tools/config.py:595]: .

File: odoo/tools/config.py [hunk 2]
```diff
@@ -506,7 +558,7 @@ def die(cond, msg):
                 self._cli_options[arg] = getattr(opt, arg)
 
         if opt.log_handler:
-            self._cli_options['log_handler'] = opt.log_handler
+            self._cli_options['log_handler'] = list(itertools.chain.from_iterable(opt.log_handler))
```
Comment (Julien00859) [inline odoo/tools/config.py:571]: .

File: odoo/tools/config.py [hunk 3]
```diff
@@ -526,59 +578,48 @@ def die(cond, msg):
...
+            *self._cli_options.get('log_handler', []),
+        ]))
+
+        self._runtime_options['init'] = dict.fromkeys(self['init'], True) or {}
+        self._runtime_options['demo'] = dict(self['init']) if not self['without_demo'] else {}
+        self._runtime_options['update'] = dict.fromkeys(self['update'], True) or {}
+        self._runtime_options['translate_modules'] = sorted(self['translate_modules'])
+
+        if 'all' in self['dev_mode']:
+            self._runtime_options['dev_mode'] = self['dev_mode'] + ['reload', 'qweb', 'xml']
```
Comment (Julien00859) [inline odoo/tools/config.py:613]: werkzeug

File: odoo/tools/config.py [hunk 4]
```diff
@@ -693,15 +792,15 @@ def load(self):
...
                             logging.WARNING,
                             "option %s reads %r in the config file at "
                             "%s but isn't a boolean option, assume %r",
-                            option.dest, 'False', self.rcfile, value
+                            option.dest, 'False', self.rcfile, value if value != [] else ''
                         )
-                elif option.type in optparse.Option.TYPE_CHECKER:
-                    value = optparse.Option.TYPE_CHECKER[option.type](option, name, value)
+                else:
+                    value = self._parse(name, value)
```
Comment (Julien00859) [inline odoo/tools/config.py:803]: .

File: odoo/tools/config.py [hunk 5]
```diff
@@ -760,6 +855,14 @@ def __getitem__(self, key):
     def root_path(self):
         return self._normalize(os.path.join(os.path.dirname(__file__), '..'))
 
+    @property
+    def addons_base_dir(self):
+        return os.path.join(self.root_path, 'addons')
+
+    @property
+    def addons_main_dir(self):
```
Comment (Julien00859) [inline odoo/tools/config.py:864]: .

File: odoo/tools/config.py [hunk 6]
```diff
@@ -615,85 +638,154 @@ def _warn_deprecated_options(self):
...
             'xmlrpc_port': 'http_port',
             'xmlrpc_interface': 'http_interface',
             'xmlrpc': 'http_enable',
         }
         p = ConfigParser.RawConfigParser()
         try:
-            p.read([self.rcfile])
+            p.read([rcfile])
             for (name,value) in p.items('options'):
                 name = outdated_options_map.get(name, name)
```
Comment (kmagusiak) [inline odoo/tools/config.py:764]: Add a warning if the option is outdated.

File: odoo/tools/config.py [hunk 7]
```diff
@@ -308,10 +354,14 @@ def _build_cli(self):
...
         # Advanced options
         group = optparse.OptionGroup(parser, "Advanced options")
-        group.add_option('--dev', dest='dev_mode', type="string", file_exportable=False,
-                         help="Enable developer mode. Param: List of options separated by comma. "
-                              "Options : all, reload, qweb, xml")
-        group.add_option('--shell-interface', dest='shell_interface', type="string", file_exportable=False,
+        group.add_option('--dev', dest='dev_mode', type='comma', my_default=[], file_exportable=False,
+                         help="Enable developer features (comma-separated list, use   "
+                              '"all" for all features). Available features:           '
+                              "- qweb: log the compiled xml with qweb errors          "
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:360]: optparse outputs --help on a fixed 79 characters width (no matter the terminal size) and it doesn't do anything clever regarding the text (nor word-wrap, nor space collapse). Using those extra spaces at the end, I make sure every option is logged on its own line. But I agree that it deserves a comment (or at the very least, this message in the commit message)

File: odoo/tools/cloc.py [hunk 8]
```diff
@@ -286,8 +286,10 @@ def count_env(self, env):
         self.count_modules(env)
         self.count_customization(env)
 
-    def count_database(self, database):
-        registry = odoo.modules.registry.Registry(config['db_name'])
+    def count_database(self, database=None):
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/cloc.py:289]: revert the change and make the `database` argument useless?

File: odoo/tools/config.py [hunk 9]
```diff
@@ -615,85 +638,154 @@ def _warn_deprecated_options(self):
...
-                raise optparse.OptionValueError("option %s: no such directory: %r" % (opt, res))
-            if not self._is_addons_path(res):
-                raise optparse.OptionValueError("option %s: the path %r is not a valid addons directory" % (opt, path))
-            ad_paths.append(res)
-
-        setattr(parser.values, option.dest, ",".join(ad_paths))
-
-    def _check_upgrade_path(self, option, opt, value, parser):
+        for path in filter(bool, map(cls._normalize, value.split(','))):
+            if not os.path.isdir(path):
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:656]: ??

File: odoo/tools/config.py (hunk 6 above)
Comment (Julien00859) [reply odoo/tools/config.py]: removed this bit altogether

File: odoo/tools/config.py [hunk 10]
```diff
@@ -21,23 +22,72 @@
...
+            'choice': optparse.check_choice,
+            'bool': cls.config._check_bool,
+            'path': cls.config._check_path,
+            'comma': cls.config._check_comma,
+            'addons_path': cls.config._check_addons_path,
+            'upgrade_path': cls.config._check_upgrade_path,
+        }
+
+    @classproperty
+    def TYPE_FORMATTER(cls):
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:51]: I don't like `isinstance(stuff, collections.abc.Iterable) and not instance(stuff, (str, bytes))`, I find it more explicit to just tell what formater to use with what option type, granted there are only 2 formater, and we surely won't need any other in the future

File: odoo/tools/config.py [hunk 11]
```diff
@@ -615,85 +638,154 @@ def _warn_deprecated_options(self):
...
 
-    def _test_enable_callback(self, option, opt, value, parser):
-        if not parser.values.test_tags:
-            parser.values.test_tags = "+standard"
+    @classmethod
+    def _check_without_demo(cls, option, opt, value):
+        return value not in ('', 'False', 'false', 'None')
+
+    @classmethod
+    def _check_bool(cls, option, opt, value):
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/tools/config.py:689]: Indeed misc is too big. Moreover, tools that import odoo could be in a module.

File: odoo/tools/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply odoo/tools/config.py:651]: You can remove that if statement as it is covered below.

**Comment by Julien00859** [comment]
The tests pass on every single commit:

```
$ git rebase master --exec "/home/julien/.pyenv/versions/odoo18/bin/python3 /home/julien/Odoo/community/odoo-bin --addons-path=~/Odoo/custo,~/Odoo/design-themes,~/Odoo/enterprise,~/Odoo/community/odoo/addons,~/Odoo/community/addons server --stop --test-tags :TestConfigManager -d odoo-ccleaner --log-level=warn"
Successfully rebased and updated refs/heads/master-ccleaner3-chainmap-juc.
```

It no longer discard undocumented options, see commit [[REF] core: ConfigCleaner(7) remove deprecated options](https://github.com/odoo/odoo/pull/185670/commits/0fd394be45cc5c2571b6ee49218b2c3b46b44404)

File: odoo/tools/config.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:659]: Just tested it, cannot do that actually, the function `_is_(upgrade|addons)_path` only work when the given path exists. I could move the test inside the function but those two `_is..._path` are also used elsewhere, I pretty much like the way it is atm :D

**Review by kmagusiak — APPROVED** [review]

**Comment by johandem** [comment]
Easy changes for our SaaS,thanks for the heads-up.

**Review by xmo-odoo — COMMENTED** [review]
@robodoo override=ci/security

File: odoo/tools/config.py [hunk 12]
```diff
@@ -615,85 +644,152 @@ def _warn_deprecated_options(self):
...
+        raise optparse.OptionValueError(
+            f"option {opt}: invalid boolean value: {value!r}"
         )
 
-    def _test_enable_callback(self, option, opt, value, parser):
-        if not parser.values.test_tags:
-            parser.values.test_tags = "+standard"
+    @classmethod
+    def _check_comma(cls, option_name, option, value):
+        return list(filter(bool, map(str.strip, value.split(','))))
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:711]: It is `[v for s in value.split(',') if v := s.strip()]` that is always incorrect.

File: odoo/tools/config.py [hunk 13]
```diff
@@ -21,23 +23,67 @@
...
+    @classproperty
+    def TYPE_CHECKER(cls):
+        return {
+            'int': lambda _option, _opt, value: int(value),
+            'float': lambda _option, _opt, value: float(value),
+            'string': lambda _option, _opt, value: str(value),
+            'choice': optparse.check_choice,
+            'bool': cls.config._check_bool,
+            'path': cls.config._check_path,
+            'comma': cls.config._check_comma,
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:45]: On my perspective I find the name short, simple and effective. What do you suggest instead?

File: odoo/tools/config.py [hunk 14]
```diff
@@ -615,85 +644,152 @@ def _warn_deprecated_options(self):
...
+            check_func = self._check_bool
+        elif option_name == 'without_demo':  # callback
+            check_func = self._check_bool  # noqa: SIM114
+        else:
+            check_func = self.parser.option_class.TYPE_CHECKER[option.type]
+        return check_func(option, option_name, value)
+
+    @classmethod
+    def _format_string(cls, value):
+        return str(value)
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py:734]: So I can use it with bool and int

File: odoo/service/server.py [hunk 15]
```diff
@@ -1189,7 +1189,7 @@ def sleep(self):
 
     def _db_list(self):
         if config['db_name']:
-            db_names = config['db_name'].split(',')
+            db_names = list(config['db_name'])
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/service/server.py:1192]: it is but I want a copy, and the odoo guideline states I should `list()` and not `.copy()`

File: odoo/tools/config.py [hunk 16]
```diff
@@ -785,10 +887,25 @@ def verify_admin_password(self, password):
...
+        return {
+            **{
+                f'source#{no}': source.get(name, empty)
+                for no, source in enumerate(self.options.maps[:-4])
+            },
+            'runtime': self._runtime_options.get(name, empty),
+            'cli': self._cli_options.get(name, empty),
+            'file': self._file_options.get(name, empty),
+            'default': self._default_options.get(name, empty),
+        }
```
[the remark this answers is not in the document]
Comment (Julien00859) [reply odoo/tools/config.py]: Moved `empty` out of the class.

This function was very useful during development. I got some type wrong, some parsing wrong, and that I could fireup a shell and use this method proved effective at debugging stuff. I would even argue that it should be a method on `ChainMap` directly.

**Comment by Julien00859** [comment]
@robodoo r+

**Pull request closed without merging**